pennylane/plugins/default_qubit.py

```diff
@@ -309,79 +309,82 @@ class DefaultQubit(QubitDevice):
-                raise ValueError("BasisState parameter and wires must be of equal length.")
-            if not self._first_operation:
-                raise DeviceError("Operation {} cannot be used after other Operations have already been applied "
-                                  "on a {} device.".format(operation.name, self.short_name))
+                A = self._get_operator_matrix(operation.name, par)
+                self._state = self.mat_vec_product(A, self._state, wires)
+
+        # store the pre-rotated state
+        self.state = self._state
+
+        # apply the circuit rotations
+        for operation in rotations:
+            wires = operation.wires
+            par = operation.parameters
+            A = self._get_operator_matrix(operation.name, par)
+            self._state = self.mat_vec_product(A, self._state, wires)
+
+    def apply_state_vector(self, wires, par):
+        input_state = np.asarray(par[0], dtype=np.complex128)
+
+        if not np.isclose(np.linalg.norm(input_state, 2), 1.0, atol=tolerance):
+            raise ValueError("Sum of amplitudes-squared does not equal one.")
+
+        n_state_vector = input_state.shape[0]
 
+        if input_state.ndim == 1 and n_state_vector == 2**len(wires):
+            # generate basis states on subset of qubits via the cartesian product
+            tuples = np.array(list(itertools.product([0, 1], repeat=len(wires))))
 
-            # get computational basis state number
-            num = int(np.dot(par[0], 2**(n - 1 - np.array(wires))))
+            # get basis states to alter on full set of qubits
+            unravelled_nums = np.zeros((2 ** len(wires), self.num_wires), dtype=int)
+            unravelled_nums[:, wires] = tuples
 
+            # get indices for which the state is changed to input state vector elements
+            nums = np.ravel_multi_index(unravelled_nums.T, [2] * self.num_wires)
             self._state = np.zeros_like(self._state)
-            self._state[num] = 1.
-            self._first_operation = False
-            return
+            self._state[nums] = input_state
+        else:
+            raise ValueError("State vector must be of length 2**wires.")
 
-        A = self._get_operator_matrix(operation.name, par)
+    def apply_basis_state(self, wires, par):
+        # length of basis state parameter
+        n_basis_state = len(par[0])
 
-        self._state = self.mat_vec_product(A, self._state, wires)
-        self._first_operation = False
+        if not set(par[0]).issubset({0, 1}):
+            raise ValueError("BasisState parameter must consist of 0 or 1 integers.")
 
-    def post_apply(self):
-        # generate computational basis samples
-        self.generate_samples()
+        if n_basis_state != len(wires):
+            raise ValueError("BasisState parameter and wires must be of equal length.")
+
+        # get computational basis state number
+        num = int(np.dot(par[0], 2**(self.num_wires - 1 - np.array(wires))))
```

Comment:
This line might be a bit lot to take in at once :sweat_smile: .
How about breaking it up as follows (or somehow similarly):
```python
basis_states = 2**(self.num_wires - 1 - np.array(wires))
num = int(np.dot(par[0], basis_states))
```